test/metabase/models/table_test.clj

```diff
@@ -177,77 +179,116 @@
+                                       :perms/create-queries :no}]
               (is (= #{table-id-1 table-id-2}
-                     (fetch-visible-ids rasta-cfg :metabase_table.id))
-                  "Clause should work with qualified keyword"))))
+                     (fetch-visible-ids user-info permission-map :id))
+                  "Clause should respect the provided permission levels")))
 
-        (testing "Non-superuser requiring only :view-data :unrestricted should see tables 1 and 2"
-          (let [rasta-view-only-cfg {:user-id       user-id
-                                     :is-superuser? false
-                                     :view-data-permission-level :unrestricted
-                                     :create-queries-permission-level :no}]
-            (is (= #{table-id-1 table-id-2}
-                   (fetch-visible-ids rasta-view-only-cfg :id))
-                "Clause should respect the provided permission levels")))
+          (testing "Non-superuser requiring only :view-data :legacy-no-self-service should see tables 1, 2, and 4"
+            (let [user-info           {:user-id       user-id
+                                       :is-superuser? false}
+                  permission-map      {:perms/view-data      :legacy-no-self-service
+                                       :perms/create-queries :no}]
+              (is (= #{table-id-1 table-id-2 table-id-4}
+                     (fetch-visible-ids user-info permission-map :id))
+                  "Clause should respect the provided permission levels")))
 
-        (testing "Non-superuser requiring :view-data :blocked should all tables since all values are more permissive "
-          (let [rasta-blocked-cfg {:user-id       user-id
-                                   :is-superuser? false
-                                   :view-data-permission-level :blocked
-                                   :create-queries-permission-level :no}]
-            (is (= #{table-id-1 table-id-2 table-id-3}
-                   (fetch-visible-ids rasta-blocked-cfg :id))
-                "Clause should filter correctly when requiring :blocked level"))))))
+          (testing "Non-superuser requiring :view-data :blocked should all tables since all values are more permissive "
+            (let [user-info           {:user-id       user-id
+                                       :is-superuser? false}
+                  permission-map      {:perms/view-data      :blocked
+                                       :perms/create-queries :no}]
+              (is (= #{table-id-1 table-id-2 table-id-3 table-id-4}
+                     (fetch-visible-ids user-info permission-map :id))
+                  "Clause should filter correctly when requiring :blocked level")))
+
+          (testing "Blocked table takes precedence over legacy-no-self-service when in any group a user is a member of"
+            (data-perms/set-table-permission! pg2 table-id-4 :perms/view-data :blocked)
```

Comment:
These tests are great/thorough but I find it very confusing to remember which table has which permissions. Part of me wonders if it'd be better (even if way more verbose) to set up permissions for reach test individually? Or name things differently? (That's hard though since there are multiple permissions for each entity, with a view-data *and* create-queries...)

Reply:
I could rename the tables to reflect the permissions set for them? `table-view-data-unrestricted-create-queries-no` doesn't seem too bad?